### PR TITLE
Dev/crypt

### PR DESCRIPTION
**Pull request by TomatorCZ**
Implementation of crypt(...) in php.
Algorithms sha256/512, md5, DES were copied from the interpreter PHP and modified(mapped) to peachpie.

### CODE REVIEW

File: src/Tests/Peachpie.Test/Program.cs [hunk 1]
```diff
@@ -36,4 +36,4 @@ static void Main(string[] args)
             }
         }
     }
-}
```
Comment (jakubmisek) [inline src/Tests/Peachpie.Test/Program.cs]: can we revert this change please :)

File: src/Peachpie.Library/Hash.cs [hunk 2]
```diff
@@ -3024,13 +3032,422 @@ public static string crypt(string str, string salt/* mandatory since 5.6 */)
...
                 // DES
-                throw new NotImplementedException();
+                return DES.Crypt(str, salt);
             }
 
             // failure
             return salt.StartsWith("*0") ? "*1" : "*0";
         }
 
+        private static string ConvertByteArrayToString(byte[] buffer)
```
Comment (jakubmisek) [inline src/Peachpie.Library/Hash.cs]: isn't it just 
```c#
Encoding.ASCII.GetString(buffer)
```
?

File: src/Peachpie.Library/Hash.cs [hunk 3]
```diff
@@ -3024,13 +3032,422 @@ public static string crypt(string str, string salt/* mandatory since 5.6 */)
...
+        private const int roundsDefault = 5000;
+        private const int roundsMin = 1000;
+        private const int roundsMax = 999999999;
+        // Table with characters for base64 transformation.
+        private const string b64t = "./0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+
+        /// <summary>
+        /// Part of cypt function in php. if sha 512 is false, sha 256 is used.
+        /// </summary>
+        private static string CryptSHA(Context ctx, string password, string salt, bool sha256)
```
Comment (jakubmisek) [inline src/Peachpie.Library/Hash.cs]: `Context ctx` is not needed

File: src/Peachpie.Library/Hash.cs [hunk 4]
```diff
@@ -3024,13 +3032,422 @@ public static string crypt(string str, string salt/* mandatory since 5.6 */)
...
+            {
+                hash[s] = (byte)itoa64[hash[s] & 0x3f];
+                s++;
+            }
+        }
+
+        /// <summary>
+        /// Part of cypt function in php. 
+        /// </summary>
+        private static string CryptMD5(Context ctx, string password, string salt)
```
Comment (jakubmisek) [inline src/Peachpie.Library/Hash.cs]: `Context ctx` is not needed.

File: src/Peachpie.Library/Hash.cs [hunk 5]
```diff
@@ -2973,12 +2973,20 @@ public static PhpArray password_get_info(string hash)
         /// One-way string hashing.
         /// </summary>
         /// <returns>Returns the hashed string or a string that is shorter than 13 characters and is guaranteed to differ from the salt on failure.</returns>
-        public static string crypt(string str, string salt/* mandatory since 5.6 */)
+        public static string crypt(Context ctxPchp, string str, string salt/* mandatory since 5.6 */)
```
Comment (jakubmisek) [inline src/Peachpie.Library/Hash.cs]: `Context ctxPchp` is not needed.

File: src/Peachpie.Library/Hash.cs [hunk 6]
```diff
@@ -2818,12 +2818,12 @@ public static PhpValue password_hash(string password, PasswordType algo, PhpArra
         /// <summary>
         /// Verifies that a password matches a hash.
         /// </summary>
-        public static bool password_verify(string password, string hash)
+        public static bool password_verify(Context ctx, string password, string hash)
```
Comment (jakubmisek) [inline src/Peachpie.Library/Hash.cs]: `Context ctx` is not needed.

File: src/Tests/Peachpie.Test/Program.cs (hunk 1 above)
Comment (TomatorCZ) [reply src/Tests/Peachpie.Test/Program.cs]: Resolved https://github.com/peachpiecompiler/peachpie/pull/593/commits/a5f7ebace380b9103c52ca3a06707f3eb80f1122

File: src/Peachpie.Library/Hash.DESHelper.cs [hunk 7]
```diff
@@ -0,0 +1,669 @@
...
+            private static uint[,] key_perm_maskl = new uint[8, 128];
+
+            private static uint[,] comp_maskl = new uint[8, 128];
+            private static uint[,] comp_maskr = new uint[8, 128];
+
+            private static uint[,] psbox = new uint[4, 256];
+
+            private static bool initialized = false;
+
+            private static readonly Object obj = new Object();
```
[the remark this answers is not in the document]
Comment (TomatorCZ) [reply src/Peachpie.Library/Hash.DESHelper.cs]: Resolved https://github.com/peachpiecompiler/peachpie/pull/593/commits/6d0d5b2be61adf4f23114024ecbc11c7c05c9304

File: src/Peachpie.Library/Hash.cs (hunk 2 above)
Comment (TomatorCZ) [reply src/Peachpie.Library/Hash.cs]: Resolved https://github.com/peachpiecompiler/peachpie/pull/593/commits/6f1865f58f6166fda85c7352d98cab2d049e5ddf

File: src/Peachpie.Library/Hash.cs (hunk 3 above)
Comment (TomatorCZ) [reply src/Peachpie.Library/Hash.cs]: Resolved https://github.com/peachpiecompiler/peachpie/pull/593/commits/6f1865f58f6166fda85c7352d98cab2d049e5ddf

File: src/Peachpie.Library/Hash.cs (hunk 4 above)
Comment (TomatorCZ) [reply src/Peachpie.Library/Hash.cs]: Resolved https://github.com/peachpiecompiler/peachpie/pull/593/commits/6f1865f58f6166fda85c7352d98cab2d049e5ddf

File: src/Peachpie.Library/Hash.cs (hunk 5 above)
Comment (TomatorCZ) [reply src/Peachpie.Library/Hash.cs]: Resolved https://github.com/peachpiecompiler/peachpie/pull/593/commits/6f1865f58f6166fda85c7352d98cab2d049e5ddf

File: src/Peachpie.Library/Hash.cs (hunk 6 above)
Comment (TomatorCZ) [reply src/Peachpie.Library/Hash.cs]: Resolved https://github.com/peachpiecompiler/peachpie/pull/593/commits/6f1865f58f6166fda85c7352d98cab2d049e5ddf

File: src/Peachpie.Library/Hash.cs [hunk 8]
```diff
@@ -2818,7 +2818,7 @@ public static PhpValue password_hash(string password, PasswordType algo, PhpArra
...
```
Comment (jakubmisek) [inline src/Peachpie.Library/Hash.cs]: `Context ctx` is not needed

**Comment by jakubmisek** [comment]
Nice! thank you @TomatorCZ
